src/Api/Client.php

```diff
@@ -9,76 +9,137 @@
 
 namespace Flarum\Api;
 
-use Exception;
-use Flarum\Foundation\ErrorHandling\JsonApiFormatter;
-use Flarum\Foundation\ErrorHandling\Registry;
 use Flarum\Http\RequestUtil;
 use Flarum\User\User;
 use Illuminate\Contracts\Container\Container;
-use InvalidArgumentException;
 use Laminas\Diactoros\ServerRequestFactory;
+use Laminas\Diactoros\Uri;
+use Laminas\Stratigility\MiddlewarePipeInterface;
 use Psr\Http\Message\ResponseInterface;
-use Psr\Http\Server\RequestHandlerInterface;
-use Throwable;
+use Psr\Http\Message\ServerRequestInterface;
 
 class Client
 {
     /**
-     * @var Container
+     * @var MiddlewarePipeInterface
      */
-    protected $container;
+    protected $pipe;
 
     /**
-     * @var Registry
+     * @var User
      */
-    protected $registry;
+    protected $actor;
+
+    /**
+     * @var ServerRequestInterface
+     */
+    protected $parent;
+
+    /**
+     * @var array
+     */
+    protected $queryParams = [];
+
+    /**
+     * @var array
+     */
+    protected $body = [];
 
     /**
      * @param Container $container
-     * @param Registry $registry
      */
-    public function __construct(Container $container, Registry $registry)
+    public function __construct(MiddlewarePipeInterface $pipe)
     {
-        $this->container = $container;
-        $this->registry = $registry;
+        $this->pipe = $pipe;
+    }
+
+    /**
+     * Set the request actor.
+     * This is not needed if a parent request is provided.
+     * It can, however, override the parent request's actor.
+     */
+    public function withActor(User $actor): Client
+    {
+        $new = clone $this;
+        $new->actor = $actor;
+
+        return $new;
+    }
+
+    public function withParentRequest(ServerRequestInterface $parent): Client
+    {
+        $new = clone $this;
+        $new->parent = $parent;
+
+        return $new;
+    }
+
+    public function withQueryParams(array $queryParams): Client
+    {
+        $new = clone $this;
+        $new->queryParams = $queryParams;
+
+        return $new;
+    }
+
+    public function withBody(array $body): Client
```

Comment:
Shouldn't we allow for text body as well with this method?

Reply:
Why?

Reply:
To cover more use cases. Though admittedly I can't think of any reason for string right now.

There's also the concern of 500 errors due to custom data... If a client sends `null` (as JSON) in the body, it should return a 400 error or just do nothing depending on the context, but here it would result in a 500 error because we try to pass down the valid JSON down to the API client which will reject it due to type-hint.